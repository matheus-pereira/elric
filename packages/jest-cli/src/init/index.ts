/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import * as path from 'path';
import chalk = require('chalk');
import * as fs from 'graceful-fs';
import prompts = require('prompts');
import {constants} from 'elric-config';
import {tryRealpath} from 'elric-util';
import {MalformedPackageJsonError, NotFoundPackageJsonError} from './errors';
import generateConfigFile from './generateConfigFile';
import modifyPackageJson from './modifyPackageJson';
import defaultQuestions, {testScriptQuestion} from './questions';
import type {ProjectPackageJson} from './types';

const {
  elric_CONFIG_BASE_NAME,
  elric_CONFIG_EXT_MJS,
  elric_CONFIG_EXT_JS,
  elric_CONFIG_EXT_TS,
  elric_CONFIG_EXT_ORDER,
  PACKAGE_JSON,
} = constants;

type PromptsResults = {
  useTypescript: boolean;
  clearMocks: boolean;
  coverage: boolean;
  coverageProvider: boolean;
  environment: boolean;
  scripts: boolean;
};

const getConfigFilename = (ext: string) => elric_CONFIG_BASE_NAME + ext;

export default async (
  rootDir: string = tryRealpath(process.cwd()),
): Promise<void> => {
  // prerequisite checks
  const projectPackageJsonPath: string = path.join(rootDir, PACKAGE_JSON);

  if (!fs.existsSync(projectPackageJsonPath)) {
    throw new NotFoundPackageJsonError(rootDir);
  }

  const questions = defaultQuestions.slice(0);
  let haselricProperty = false;
  let projectPackageJson: ProjectPackageJson;

  try {
    projectPackageJson = JSON.parse(
      fs.readFileSync(projectPackageJsonPath, 'utf-8'),
    );
  } catch {
    throw new MalformedPackageJsonError(projectPackageJsonPath);
  }

  if (projectPackageJson.elric) {
    haselricProperty = true;
  }

  const existingelricConfigExt = elric_CONFIG_EXT_ORDER.find(ext =>
    fs.existsSync(path.join(rootDir, getConfigFilename(ext))),
  );

  if (haselricProperty || existingelricConfigExt) {
    const result: {continue: boolean} = await prompts({
      initial: true,
      message:
        'It seems that you already have a elric configuration, do you want to override it?',
      name: 'continue',
      type: 'confirm',
    });

    if (!result.continue) {
      console.log();
      console.log('Aborting...');
      return;
    }
  }

  // Add test script installation only if needed
  if (
    !projectPackageJson.scripts ||
    projectPackageJson.scripts.test !== 'elric'
  ) {
    questions.unshift(testScriptQuestion);
  }

  // Start the init process
  console.log();
  console.log(
    chalk.underline(
      `The following questions will help elric to create a suitable configuration for your project\n`,
    ),
  );

  let promptAborted = false;

  // @ts-expect-error: Return type cannot be object - faulty typings
  const results: PromptsResults = await prompts(questions, {
    onCancel: () => {
      promptAborted = true;
    },
  });

  if (promptAborted) {
    console.log();
    console.log('Aborting...');
    return;
  }

  // Determine if elric should use JS or TS for the config file
  const elricConfigFileExt = results.useTypescript
    ? elric_CONFIG_EXT_TS
    : projectPackageJson.type === 'module'
    ? elric_CONFIG_EXT_MJS
    : elric_CONFIG_EXT_JS;

  // Determine elric config path
  const elricConfigPath = existingelricConfigExt
    ? getConfigFilename(existingelricConfigExt)
    : path.join(rootDir, getConfigFilename(elricConfigFileExt));

  const shouldModifyScripts = results.scripts;

  if (shouldModifyScripts || haselricProperty) {
    const modifiedPackageJson = modifyPackageJson({
      projectPackageJson,
      shouldModifyScripts,
    });

    fs.writeFileSync(projectPackageJsonPath, modifiedPackageJson);

    console.log('');
    console.log(`‚úèÔ∏è  Modified ${chalk.cyan(projectPackageJsonPath)}`);
  }

  const generatedConfig = generateConfigFile(
    results,
    projectPackageJson.type === 'module' ||
      elricConfigPath.endsWith(elric_CONFIG_EXT_MJS),
  );

  fs.writeFileSync(elricConfigPath, generatedConfig);

  console.log('');
  console.log(
    `üìù  Configuration file created at ${chalk.cyan(elricConfigPath)}`,
  );
};
