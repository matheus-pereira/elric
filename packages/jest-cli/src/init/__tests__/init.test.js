/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

/* eslint-disable no-eval */
import * as path from 'path';
import * as fs from 'graceful-fs';
import prompts from 'prompts';
import {constants} from 'elric-config';
import init from '../';

const {elric_CONFIG_EXT_ORDER} = constants;

elric.mock('prompts');
elric.mock(
  '../../../../elric-config/build/getCacheDirectory',
  () => () => '/tmp/elric',
);
elric.mock('path', () => ({...elric.requireActual('path'), sep: '/'}));
elric.mock('graceful-fs', () => ({
  ...elric.requireActual('fs'),
  writeFileSync: elric.fn(),
}));

const resolveFromFixture = relativePath =>
  path.resolve(__dirname, 'fixtures', relativePath);

const consoleLog = console.log;

describe('init', () => {
  beforeEach(() => {
    console.log = elric.fn();
  });

  afterEach(() => {
    elric.clearAllMocks();
    console.log = consoleLog;
  });

  describe('project with package.json and no elric config', () => {
    describe('all questions answered with answer: "No"', () => {
      it('should return the default configuration (an empty config)', async () => {
        prompts.mockReturnValueOnce({});

        await init(resolveFromFixture('only-package-json'));

        const writtenelricConfig = fs.writeFileSync.mock.calls[0][1];

        expect(writtenelricConfig).toMatchSnapshot();

        const evaluatedConfig = eval(writtenelricConfig);

        expect(evaluatedConfig).toEqual({});
      });

      it('should generate empty config with mjs extension', async () => {
        prompts.mockReturnValueOnce({});

        await init(resolveFromFixture('type-module'));

        const writtenelricConfigFilename = fs.writeFileSync.mock.calls[0][0];
        const writtenelricConfig = fs.writeFileSync.mock.calls[0][1];

        expect(path.basename(writtenelricConfigFilename)).toBe(
          'elric.config.mjs',
        );

        expect(typeof writtenelricConfig).toBe('string');
        expect(writtenelricConfig.split('\n')[5]).toBe('export default {');
      });
    });

    describe('some questions answered with answer: "Yes"', () => {
      it('should create configuration for {clearMocks: true}', async () => {
        prompts.mockReturnValueOnce({clearMocks: true});

        await init(resolveFromFixture('only-package-json'));

        const writtenelricConfig = fs.writeFileSync.mock.calls[0][1];
        const evaluatedConfig = eval(writtenelricConfig);

        expect(evaluatedConfig).toEqual({clearMocks: true});
      });

      it('should create configuration for {coverage: true}', async () => {
        prompts.mockReturnValueOnce({coverage: true});

        await init(resolveFromFixture('only-package-json'));

        const writtenelricConfig = fs.writeFileSync.mock.calls[0][1];
        const evaluatedConfig = eval(writtenelricConfig);

        expect(evaluatedConfig).toEqual({
          collectCoverage: true,
          coverageDirectory: 'coverage',
        });
      });

      it('should create configuration for {coverageProvider: "babel"}', async () => {
        prompts.mockReturnValueOnce({coverageProvider: 'babel'});

        await init(resolveFromFixture('only-package-json'));

        const writtenelricConfig = fs.writeFileSync.mock.calls[0][1];
        const evaluatedConfig = eval(writtenelricConfig);
        // should modify when the default coverageProvider will be changed to "v8"
        expect(evaluatedConfig).toEqual({});
      });

      it('should create configuration for {coverageProvider: "v8"}', async () => {
        prompts.mockReturnValueOnce({coverageProvider: 'v8'});

        await init(resolveFromFixture('only-package-json'));

        const writtenelricConfig = fs.writeFileSync.mock.calls[0][1];
        const evaluatedConfig = eval(writtenelricConfig);
        // should modify when the default coverageProvider will be changed to "v8"
        expect(evaluatedConfig).toEqual({coverageProvider: 'v8'});
      });

      it('should create configuration for {environment: "jsdom"}', async () => {
        prompts.mockReturnValueOnce({environment: 'jsdom'});

        await init(resolveFromFixture('only-package-json'));

        const writtenelricConfig = fs.writeFileSync.mock.calls[0][1];
        const evaluatedConfig = eval(writtenelricConfig);
        expect(evaluatedConfig).toEqual({testEnvironment: 'jsdom'});
      });

      it('should create configuration for {environment: "node"}', async () => {
        prompts.mockReturnValueOnce({environment: 'node'});

        await init(resolveFromFixture('only-package-json'));

        const writtenelricConfig = fs.writeFileSync.mock.calls[0][1];
        const evaluatedConfig = eval(writtenelricConfig);
        expect(evaluatedConfig).toEqual({});
      });

      it('should create package.json with configured test command when {scripts: true}', async () => {
        prompts.mockReturnValueOnce({scripts: true});

        await init(resolveFromFixture('only-package-json'));

        const writtenPackageJson = fs.writeFileSync.mock.calls[0][1];

        expect(writtenPackageJson).toMatchSnapshot();
        expect(JSON.parse(writtenPackageJson).scripts.test).toEqual('elric');
      });
    });
  });

  describe('no package json', () => {
    it('should throw an error if there is no package.json file', async () => {
      expect.assertions(1);

      try {
        await init(resolveFromFixture('no-package-json'));
      } catch (error) {
        expect(error.message).toMatch(
          'Could not find a "package.json" file in',
        );
      }
    });
  });

  describe.each(elric_CONFIG_EXT_ORDER.map(e => e.substring(1)))(
    'has-elric-config-file-%s',
    extension => {
      describe('ask the user whether to override config or not', () => {
        it('user answered with "Yes"', async () => {
          prompts.mockReturnValueOnce({continue: true}).mockReturnValueOnce({});

          await init(resolveFromFixture(`has-elric-config-file-${extension}`));

          expect(prompts.mock.calls[0][0]).toMatchSnapshot();

          const elricConfigFileName = fs.writeFileSync.mock.calls[0][0];
          const writtenelricConfig = fs.writeFileSync.mock.calls[0][1];

          expect(elricConfigFileName).toBe(`elric.config.${extension}`);
          expect(writtenelricConfig).toBeDefined();
        });

        it('user answered with "No"', async () => {
          prompts.mockReturnValueOnce({continue: false});

          await init(resolveFromFixture(`has-elric-config-file-${extension}`));
          // return after first prompt
          expect(prompts).toHaveBeenCalledTimes(1);
        });
      });
    },
  );

  describe('project using elric.config.ts', () => {
    describe('ask the user whether he wants to use Typescript or not', () => {
      it('user answered with "Yes"', async () => {
        prompts.mockReturnValueOnce({useTypescript: true});

        await init(resolveFromFixture('test-generated-elric-config-ts'));

        expect(prompts.mock.calls[0][0]).toMatchSnapshot();

        const elricConfigFileName = fs.writeFileSync.mock.calls[0][0];
        const writtenelricConfig = fs.writeFileSync.mock.calls[0][1];

        expect(path.basename(elricConfigFileName)).toBe('elric.config.ts');
        expect(writtenelricConfig.split('\n')[5]).toBe('export default {');
      });

      it('user answered with "No"', async () => {
        prompts.mockReturnValueOnce({useTypescript: false});

        await init(resolveFromFixture('test-generated-elric-config-ts'));

        const elricConfigFileName = fs.writeFileSync.mock.calls[0][0];

        expect(path.basename(elricConfigFileName)).not.toBe('elric.config.ts');
      });
    });
  });

  describe('has elric config in package.json', () => {
    it('should ask the user whether to override config or not', async () => {
      prompts.mockReturnValueOnce({continue: true}).mockReturnValueOnce({});

      await init(resolveFromFixture('has-elric-config-in-package-json'));

      expect(prompts.mock.calls[0][0]).toMatchSnapshot();

      const writtenelricConfig = fs.writeFileSync.mock.calls[0][1];

      expect(writtenelricConfig).toBeDefined();
    });
  });

  describe('already has "elric" in packageJson.scripts.test', () => {
    it('should not ask "test script question"', async () => {
      prompts.mockReturnValueOnce({});

      await init(resolveFromFixture('test-script-configured'));

      const questionsNames = prompts.mock.calls[0][0].map(
        question => question.name,
      );

      expect(questionsNames).not.toContain('scripts');
    });
  });
});
