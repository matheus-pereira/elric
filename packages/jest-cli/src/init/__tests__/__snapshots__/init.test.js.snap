// elric Snapshot v1, https://goo.gl/fbAQLP

exports[`init has elric config in package.json should ask the user whether to override config or not 1`] = `
Object {
  "initial": true,
  "message": "It seems that you already have a elric configuration, do you want to override it?",
  "name": "continue",
  "type": "confirm",
}
`;

exports[`init has-elric-config-file-cjs ask the user whether to override config or not user answered with "Yes" 1`] = `
Object {
  "initial": true,
  "message": "It seems that you already have a elric configuration, do you want to override it?",
  "name": "continue",
  "type": "confirm",
}
`;

exports[`init has-elric-config-file-js ask the user whether to override config or not user answered with "Yes" 1`] = `
Object {
  "initial": true,
  "message": "It seems that you already have a elric configuration, do you want to override it?",
  "name": "continue",
  "type": "confirm",
}
`;

exports[`init has-elric-config-file-json ask the user whether to override config or not user answered with "Yes" 1`] = `
Object {
  "initial": true,
  "message": "It seems that you already have a elric configuration, do you want to override it?",
  "name": "continue",
  "type": "confirm",
}
`;

exports[`init has-elric-config-file-mjs ask the user whether to override config or not user answered with "Yes" 1`] = `
Object {
  "initial": true,
  "message": "It seems that you already have a elric configuration, do you want to override it?",
  "name": "continue",
  "type": "confirm",
}
`;

exports[`init has-elric-config-file-ts ask the user whether to override config or not user answered with "Yes" 1`] = `
Object {
  "initial": true,
  "message": "It seems that you already have a elric configuration, do you want to override it?",
  "name": "continue",
  "type": "confirm",
}
`;

exports[`init project using elric.config.ts ask the user whether he wants to use Typescript or not user answered with "Yes" 1`] = `
Array [
  Object {
    "initial": true,
    "message": "Would you like to use elric when running \\"test\\" script in \\"package.json\\"?",
    "name": "scripts",
    "type": "confirm",
  },
  Object {
    "initial": false,
    "message": "Would you like to use Typescript for the configuration file?",
    "name": "useTypescript",
    "type": "confirm",
  },
  Object {
    "choices": Array [
      Object {
        "title": "node",
        "value": "node",
      },
      Object {
        "title": "jsdom (browser-like)",
        "value": "jsdom",
      },
    ],
    "initial": 0,
    "message": "Choose the test environment that will be used for testing",
    "name": "environment",
    "type": "select",
  },
  Object {
    "initial": false,
    "message": "Do you want elric to add coverage reports?",
    "name": "coverage",
    "type": "confirm",
  },
  Object {
    "choices": Array [
      Object {
        "title": "v8",
        "value": "v8",
      },
      Object {
        "title": "babel",
        "value": "babel",
      },
    ],
    "initial": 0,
    "message": "Which provider should be used to instrument code for coverage?",
    "name": "coverageProvider",
    "type": "select",
  },
  Object {
    "initial": false,
    "message": "Automatically clear mock calls, instances and results before every test?",
    "name": "clearMocks",
    "type": "confirm",
  },
]
`;

exports[`init project with package.json and no elric config all questions answered with answer: "No" should return the default configuration (an empty config) 1`] = `
"/*
 * For a detailed explanation regarding each configuration property, visit:
 * https://elricjs.io/docs/configuration
 */

module.exports = {
  // All imported modules in your tests should be mocked automatically
  // automock: false,

  // Stop running tests after \`n\` failures
  // bail: 0,

  // The directory where elric should store its cached dependency information
  // cacheDirectory: \\"/tmp/elric\\",

  // Automatically clear mock calls, instances and results before every test
  // clearMocks: false,

  // Indicates whether the coverage information should be collected while executing the test
  // collectCoverage: false,

  // An array of glob patterns indicating a set of files for which coverage information should be collected
  // collectCoverageFrom: undefined,

  // The directory where elric should output its coverage files
  // coverageDirectory: undefined,

  // An array of regexp pattern strings used to skip coverage collection
  // coveragePathIgnorePatterns: [
  //   \\"/node_modules/\\"
  // ],

  // Indicates which provider should be used to instrument code for coverage
  // coverageProvider: \\"babel\\",

  // A list of reporter names that elric uses when writing coverage reports
  // coverageReporters: [
  //   \\"json\\",
  //   \\"text\\",
  //   \\"lcov\\",
  //   \\"clover\\"
  // ],

  // An object that configures minimum threshold enforcement for coverage results
  // coverageThreshold: undefined,

  // A path to a custom dependency extractor
  // dependencyExtractor: undefined,

  // Make calling deprecated APIs throw helpful error messages
  // errorOnDeprecated: false,

  // Force coverage collection from ignored files using an array of glob patterns
  // forceCoverageMatch: [],

  // A path to a module which exports an async function that is triggered once before all test suites
  // globalSetup: undefined,

  // A path to a module which exports an async function that is triggered once after all test suites
  // globalTeardown: undefined,

  // A set of global variables that need to be available in all test environments
  // globals: {},

  // The maximum amount of workers used to run your tests. Can be specified as % or a number. E.g. maxWorkers: 10% will use 10% of your CPU amount + 1 as the maximum worker number. maxWorkers: 2 will use a maximum of 2 workers.
  // maxWorkers: \\"50%\\",

  // An array of directory names to be searched recursively up from the requiring module's location
  // moduleDirectories: [
  //   \\"node_modules\\"
  // ],

  // An array of file extensions your modules use
  // moduleFileExtensions: [
  //   \\"js\\",
  //   \\"jsx\\",
  //   \\"ts\\",
  //   \\"tsx\\",
  //   \\"json\\",
  //   \\"node\\"
  // ],

  // A map from regular expressions to module names or to arrays of module names that allow to stub out resources with a single module
  // moduleNameMapper: {},

  // An array of regexp pattern strings, matched against all module paths before considered 'visible' to the module loader
  // modulePathIgnorePatterns: [],

  // Activates notifications for test results
  // notify: false,

  // An enum that specifies notification mode. Requires { notify: true }
  // notifyMode: \\"failure-change\\",

  // A preset that is used as a base for elric's configuration
  // preset: undefined,

  // Run tests from one or more projects
  // projects: undefined,

  // Use this configuration option to add custom reporters to elric
  // reporters: undefined,

  // Automatically reset mock state before every test
  // resetMocks: false,

  // Reset the module registry before running each individual test
  // resetModules: false,

  // A path to a custom resolver
  // resolver: undefined,

  // Automatically restore mock state and implementation before every test
  // restoreMocks: false,

  // The root directory that elric should scan for tests and modules within
  // rootDir: undefined,

  // A list of paths to directories that elric should use to search for files in
  // roots: [
  //   \\"<rootDir>\\"
  // ],

  // Allows you to use a custom runner instead of elric's default test runner
  // runner: \\"elric-runner\\",

  // The paths to modules that run some code to configure or set up the testing environment before each test
  // setupFiles: [],

  // A list of paths to modules that run some code to configure or set up the testing framework before each test
  // setupFilesAfterEnv: [],

  // The number of seconds after which a test is considered as slow and reported as such in the results.
  // slowTestThreshold: 5,

  // A list of paths to snapshot serializer modules elric should use for snapshot testing
  // snapshotSerializers: [],

  // The test environment that will be used for testing
  // testEnvironment: \\"elric-environment-node\\",

  // Options that will be passed to the testEnvironment
  // testEnvironmentOptions: {},

  // Adds a location field to test results
  // testLocationInResults: false,

  // The glob patterns elric uses to detect test files
  // testMatch: [
  //   \\"**/__tests__/**/*.[jt]s?(x)\\",
  //   \\"**/?(*.)+(spec|test).[tj]s?(x)\\"
  // ],

  // An array of regexp pattern strings that are matched against all test paths, matched tests are skipped
  // testPathIgnorePatterns: [
  //   \\"/node_modules/\\"
  // ],

  // The regexp pattern or array of patterns that elric uses to detect test files
  // testRegex: [],

  // This option allows the use of a custom results processor
  // testResultsProcessor: undefined,

  // This option allows use of a custom test runner
  // testRunner: \\"elric-circus/runner\\",

  // This option sets the URL for the jsdom environment. It is reflected in properties such as location.href
  // testURL: \\"http://localhost\\",

  // Setting this value to \\"fake\\" allows the use of fake timers for functions such as \\"setTimeout\\"
  // timers: \\"real\\",

  // A map from regular expressions to paths to transformers
  // transform: undefined,

  // An array of regexp pattern strings that are matched against all source file paths, matched files will skip transformation
  // transformIgnorePatterns: [
  //   \\"/node_modules/\\",
  //   \\"\\\\\\\\.pnp\\\\\\\\.[^\\\\\\\\/]+$\\"
  // ],

  // An array of regexp pattern strings that are matched against all modules before the module loader will automatically return a mock for them
  // unmockedModulePathPatterns: undefined,

  // Indicates whether each individual test should be reported during the run
  // verbose: undefined,

  // An array of regexp patterns that are matched against all source file paths before re-running tests in watch mode
  // watchPathIgnorePatterns: [],

  // Whether to use watchman for file crawling
  // watchman: true,
};
"
`;

exports[`init project with package.json and no elric config some questions answered with answer: "Yes" should create package.json with configured test command when {scripts: true} 1`] = `
"{
  \\"name\\": \\"only_package_json\\",
  \\"scripts\\": {
    \\"test\\": \\"elric\\"
  }
}
"
`;
